src/renderer/mesh_loader.zig

```diff
@@ -1,18 +1,192 @@
 const std = @import("std");
-const assert = std.debug.assert;
 const zmesh = @import("zmesh");
 const zm = @import("zmath");
 
-const IndexType = @import("renderer_types.zig").IndexType;
-const Vertex = @import("renderer_types.zig").Vertex;
-const Mesh = @import("renderer_types.zig").Mesh;
+const rt = @import("renderer_types.zig");
+
+const assert = std.debug.assert;
+const zcgltf = zmesh.io.zcgltf;
+
+pub fn parseMeshPrimitive(
+    primitive: *zcgltf.Primitive,
+    meshes_indices: *std.ArrayList(rt.IndexType),
+    meshes_vertices: *std.ArrayList(rt.Vertex),
+    arena: std.mem.Allocator,
+) !rt.Mesh {
+    const num_vertices: u32 = @as(u32, @intCast(primitive.attributes[0].data.count));
+    const num_indices: u32 = @as(u32, @intCast(primitive.indices.?.count));
+
+    var indices = std.ArrayList(rt.IndexType).init(arena);
+    var positions = std.ArrayList([3]f32).init(arena);
+    var normals = std.ArrayList([3]f32).init(arena);
+    var tangents = std.ArrayList([4]f32).init(arena);
+    var uvs = std.ArrayList([2]f32).init(arena);
+    defer indices.deinit();
+    defer positions.deinit();
+    defer normals.deinit();
+    defer tangents.deinit();
+    defer uvs.deinit();
+
+    // Indices.
+    {
+        try indices.ensureTotalCapacity(indices.items.len + num_indices);
+
+        const accessor = primitive.indices.?;
+        const buffer_view = accessor.buffer_view.?;
+
+        assert(accessor.stride == buffer_view.stride or buffer_view.stride == 0);
+        assert(accessor.stride * accessor.count == buffer_view.size);
+        assert(buffer_view.buffer.data != null);
+
+        const data_addr = @as([*]const u8, @ptrCast(buffer_view.buffer.data)) +
+            accessor.offset + buffer_view.offset;
+
+        if (accessor.stride == 1) {
+            assert(accessor.component_type == .r_8u);
+            const src = @as([*]const u8, @ptrCast(data_addr));
+            var i: u32 = 0;
+            while (i < num_indices) : (i += 1) {
+                indices.appendAssumeCapacity(src[i]);
+            }
+        } else if (accessor.stride == 2) {
+            assert(accessor.component_type == .r_16u);
+            const src = @as([*]const u16, @ptrCast(@alignCast(data_addr)));
+            var i: u32 = 0;
+            while (i < num_indices) : (i += 1) {
```

Comment:
This could be a regular for loop now right?

Reply:
It looks like we can't use a for loop since `src` doesn't have an upper bound
```
type '[*]const u8' has no upper bound
```

Reply:
Ah but you could do `for (0..num_indices) |i|` right?

Reply:
oh yes, I can do that